deepchem/models/torch_models/grover.py

```diff
@@ -1,94 +1,397 @@
+
+    def _prepare_batch(self):
+        if self.task == 'pretraining':
+            return self._prepare_batch_for_pretraining()
+        elif self.task == 'finetuning':
+            return self._prepare_batch_for_finetuning()
+
+    def _prepare_batch_for_pretraining(self, batchgraph: Sequence[MolGraph],
+                                       atom_vocab, bond_vocab):
+        """
         Parameters
         ----------
-        graph_batch: List[torch.Tensor]
-            A list containing f_atoms, f_bonds, a2b, b2a, b2revb, a_scope, b_scope, a2a
+        smiles: List[str]
+            A list of smiles strings
+        atom_vocab: MolVocab
+            atom vocabulary
+        bond_vocab: MolVocab
+            bond vocabulary
         """
-        output = self.encoders(graph_batch)
-        if self.atom_embedding_output_type == 'atom':
-            return {
-                "atom_from_atom": output[0],
-                "atom_from_bond": output[1],
-                "bond_from_atom": None,
-                "bond_from_bond": None
-            }  # atom_from_atom, atom_from_bond
-        elif self.atom_embedding_output_type == 'bond':
-            return {
-                "atom_from_atom": None,
-                "atom_from_bond": None,
-                "bond_from_atom": output[0],
-                "bond_from_bond": output[1]
-            }  # bond_from_atom, bond_from_bond
-        elif self.atom_embedding_output_type == "both":
-            return {
-                "atom_from_atom": output[0][0],
-                "bond_from_atom": output[0][1],
-                "atom_from_bond": output[1][0],
-                "bond_from_bond": output[1][1]
+        batchgraph = BatchMolGraph(batchgraph)
+        fgroup_label = batchgraph.fg_labels
+        smiles_batch = batchgraph.smiles_batch
+
+        batchgraph = batchgraph.get_components()
+
+        atom_vocab_label = torch.Tensor(
+            GroverPretrain.atom_random_mask(atom_vocab, smiles_batch)).long()
+        bond_vocab_label = torch.Tensor(
+            GroverPretrain.bond_random_mask(bond_vocab, smiles_batch)).long()
+        res = {
+            "graph_input": batchgraph,
+            "targets": {
+                "av_task": atom_vocab_label,
+                "bv_task": bond_vocab_label,
+                "fg_task": fgroup_label
             }
+        }
+        return res
+
+    @staticmethod
+    def _prepare_batch_for_finetuning(self, batchgraph: Sequence[MolGraph],
+                                      targets):
+        """
+        Parameters
+        ---------
+        batchgraph: List[MolGraph]
+            list of smiles strings
+        targets: List[Sequence]
+            targets for the batch
+        """
+        # TODO Replace BatchMolGraph with _grover_batch_graph method
+        batchgraph = BatchMolGraph(batchgraph)
+        targets = torch.FloatTensor(np.array(targets))
+        return batchgraph, targets
+
+    def _get_pretraining_loss_func(preds,
```

Comment:
This loss function should be converted to a function and moved to the section of deepchem losses since other models may want to reuse this loss as a general mathematical operation